HPCC-15191 Extend TxSummary functionality to support timing milestones

timothyklemm opened the pull request:
Add timer class to track total elapsed time reported in one or more steps.

Update TxSummary to manage a collection of timer instances, referenced by name,
Extend the summary API to permit lookup of, and updates to, managed timers.
Add managed timer values to the serialized summary.

Extend the IEspContext interface to provide pass-through functions comparable
to its existing "trace summary" interface.

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community_6.0.0-beta1.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

timothyklemm commented:
@rpastrana Take a look.

richardkchapman commented:
@ghalliday Can you look too? This feels like it may be re-inventing some things we have elsewhere.

rpastrana commented on esp/platform/espcontext.cpp:
```diff
@@ -454,6 +454,15 @@ class CEspContext : public CInterface, implements IEspContext
+    virtual void addTraceSummaryCumulativeTime(const char* name, unsigned __int64 time)
+    {
+        if (m_txSummary)
+            m_txSummary->updateTimer(name, time);
```
is name=NULL safe?

timothyklemm replied:
NULL and empty names will be rejected by updateTimer, which relies on getTimer to validate the input. The context doesn't validate the input because I wanted the object that imposes the requirements to have sole responsibility for validation.

ghalliday replied:
How frequently do you plan to call this function?  Is a hash table lookup each time likely to be a significant cost?

timothyklemm replied:
This method was added for consistency - the context can add key value pairs, so it seemed appropriate to allow timer updates. Common usage has been to get and retain timer instances rather than using this method.

rpastrana commented on esp/platform/txsummary.cpp:
```diff
@@ -21,10 +21,16 @@
 #define VALIDATE_KEY(k) if (!(k) || !(*k)) return false
 #define MATCH_KEY       [&](const Entry& entry) { return stricmp(entry.key.str(), key) == 0; }
 
+bool operator < (const StringBuffer& a, const StringBuffer& b)
```
why do we need to provide this operator definition? and if it's needed, do we also need < >=, etc?

timothyklemm replied:
std::map is an ordered collection, with the default comparison implemented by std::less. std::less uses operator < for its comparisons. To use StringBuffer as a key in std::map requires a comparison for less than and no other comparison.

rpastrana commented on esp/platform/txsummary.cpp:
```diff
@@ -81,8 +97,41 @@ bool CTxSummary::set(const char* key, const char* value)
+    if (!name || !*name)
+        return NULL;
+
+    synchronized block(m_sync);
```
I like to make these boundaries as tight as possible, do we need to protect the instantiation of key?

timothyklemm replied:
No

rpastrana commented on esp/platform/txsummary.cpp:
```diff
@@ -90,6 +139,11 @@ void CTxSummary::serialize(StringBuffer& buffer) const
+
+    for (const std::pair<StringBuffer, CumulativeTimer*>& entry : m_timers)
+    {
+        buffer.appendf("%s=%" I64F "ums;", entry.first.str(), entry.second->getTotalMillis());
```
the string format is showing up as "%s=%" I64F "ums;" on github, please make sure this is correct.

timothyklemm replied:
This is a bit strange, but correct. The format token for an unsigned 64 bit integer differs between Linux ("%llu") and Windows ("%I64u"). The syntax in question allows the preprocessor to assemble the correct form of the format string.

rpastrana commented:
@timothyklemm reviewed. Left several comments.

rpastrana commented:
@richardkchapman perhaps we can look into the potentially redundant feature, which feature/component do you suspect? Thanks.

rpastrana commented:
@timothyklemm could jdebug's TimeSection class be used in this feature?

timothyklemm commented:
Neither TimeSection nor MTimeSection are suitable. While they do measure elapsed time, they do so in a specific way that is incompatible with the CTxSummary requirements.

The ITimeReporter interface, also in jdebug, does declare somewhat similar functionality. A concrete implementation can be defined to be used with CTxSummary, but I don't know if it is worthwhile. My impression of the interface relative to the CTxSummary requirements is that the instance would be encapsulated within the summary with no addition functionality exposed. It wouldn't replace the CumulativeTimer class, because I see that as the the implementation of the section. With respect to sections, the expectation of the summary is that any section can be linked by other objects and updated directly without repeatedly looking it up in the owning object. Again, although there are similarities I don't think interface gains anything in this case.

ghalliday commented on esp/platform/cumulativetimer.hpp:
```diff
@@ -0,0 +1,106 @@
+    inline unsigned __int64 getTotalMillis() const { return mTotalTime; }
+    inline void reset() { mTotalTime = 0; }
+
+    IMPLEMENT_IINTERFACE;
```
minor: not needed.  Only needed if the class implements an interface

ghalliday commented on esp/platform/cumulativetimer.hpp:
```diff
@@ -0,0 +1,106 @@
+        CumulativeTimer*    mTimer;
+
+    private:
+        Scope(const Scope& t);
```
minor style: could use the c++11 = delete syntax.

ghalliday commented on esp/platform/txsummary.hpp:
```diff
@@ -78,9 +91,12 @@ class CTxSummary : extends CInterface
     };
 
     using Entries = std::list<Entry>;
+    using Timers = std::map<StringBuffer, CumulativeTimer*>;
```
I would have expected something like std::map<StringAttr, Linked<CumulativeTime> >

- StringAttr is much smaller than a StringBuffer
- using a wrapper class means that the mapped objects should be cleaned up correctly

ghalliday commented on esp/platform/txsummary.cpp:
```diff
@@ -81,8 +97,41 @@ bool CTxSummary::set(const char* key, const char* value)
     return true;
 }
 
+CumulativeTimer* CTxSummary::getTimer(const char* name)
```
Our convention is normally to call the function queryTimer if it does not return a linked object.

ghalliday commented on esp/platform/cumulativetimer.hpp:
```diff
@@ -0,0 +1,106 @@
+
+    void add(unsigned __int64 delta)
+    {
+        if (mNestingDepth == 0)
```
Have you thought about thread safety:

- If two threads have called getTimer() and then call add() the value will potentially be lost

- It two threads have both entered a Scope on the same timer then one of the timings may well be lost.  If there is always a thread with an active scope for the timer then all timings will be lost.

timothyklemm replied:
Agreed. This code was not written by me and is in production, so I assumed thread safety had been addressed. I learned last night that thread safety is assumed based on calling conventions. Will fix.

ghalliday commented on esp/platform/cumulativetimer.hpp:
```diff
@@ -0,0 +1,106 @@
+private:
+    friend struct CumulativeTimer::Scope;
+
+    inline void incNesting() { ++mNestingDepth; }
```
not thread safe, but also has other threading problems (see above)

ghalliday commented on esp/platform/txsummary.hpp:
```diff
@@ -78,9 +91,12 @@ class CTxSummary : extends CInterface
 
-    Entries   m_entries;
-    unsigned  m_creationTime;
+    mutable Monitor m_sync;
```
minor: The platform code generally used a CriticalSection/CriticalBlock for a mutex which has no timeout.  It is lighter weight (especially on windows) than a mutex.

timothyklemm replied:
My preference for mutex, even within a single process, is based on recent experience as a Windows developer. We substituted mutex for critical section in all of our projects after observing unexplained deadlocks with critical sections. If critical sections are working reliably, I'll make the change.

timothyklemm commented:
@rpastrana and @ghalliday I've updated based on comments. Please check.

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community_6.0.0-beta1.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

richardkchapman commented:
Needs rebasing for a clean merge

ghalliday commented on esp/platform/cumulativetimer.hpp:
```diff
@@ -0,0 +1,110 @@
+
+    void add(unsigned __int64 delta)
+    {
+        CriticalBlock b(mSync);
```
You should probably be using atomic operations here, rather than criticalblocks.  e.g., std::atomic

timothyklemm replied:
I considered it, and chose the CriticalBlock instead. I want to ensure that the condition on mNestingDepth remains true when mTotalTime is changed. Both in add, and in decNesting which combines what was two separate timer methods called by Scope. @ghalliday, will atomic operations accomplish this?

ghalliday replied:
I think you need to clarify the specificiation - whether a timer instance can be updated from multiple threads or not.

If the answer is yes, then the idea behind the Scope class doesn't work - because timers on one thread will prevent timers from other threads being updated.

If the answer is no (or that we don't care about small race conditions), then it is much more efficient to remove the critical section.

Whichever it is should be included as a comment before the class to make it clear.

ghalliday commented:
@timothyklemm please can you clarify what you mean by "I learned last night that thread safety is assumed based on calling conventions."

In particular, does the intended design allow for multiple threads to be accessing the same timer at the same time.  Also, are multiple threads allowed to have active Scope object instances at the same time?

It can you clarify how time critical this code is.  If it is called once per esp call then efficiency is not a big concern - and using a critical section to protect a count is probably ok, although not ideal.  If it is called many times then it calls for more consideration of efficiency.

Thanks,
Gavin

timothyklemm commented:
@ghalliday in the original environment, timers are accessed through thread local storage by way of a pointer to the EspContext added to TLS. If a timer accessed this way is never passed to code executing in another thread, there is no need for explicit synchronization. I agreed with the comments about missing synchronization because it seems unreasonable to assume that this is the only way in which timers will be used.

As we use timers now, in our security manager plugin, I have observed multiple timers being created per esp call, with each timer updated multiple times.

timothyklemm commented:
@richardkchapman I will rebase, but want to wait until everything else is approved so I don't have to do it multiple times. I checked yesterday, and there was only one conflict - two new, unrelated, methods added to CEspContext immediately before my rename of getTxSummary to queryTxSummary. It is a trivial resolution.

richardkchapman commented:
@ghalliday I think this one is in your court at the moment?

rpastrana commented:
@timothyklemm latest squashed commit seems to have addressed all my concerns

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community_6.0.0-beta1.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

rpastrana commented:
@ghalliday back on your court...

ghalliday commented on esp/platform/txsummary.cpp:
```diff
@@ -81,8 +94,40 @@ bool CTxSummary::set(const char* key, const char* value)
+bool CTxSummary::updateTimer(const char* name, unsigned long long delta)
+{
+    Owned<CumulativeTimer> timer;
+    CriticalBlock block(m_sync);
```
this critical block shouldn't be needed.

rpastrana replied:
@timothyklemm can you address @ghalliday 's comment? Either reply, or implement his proposed change.

ghalliday commented:
@timothyklemm one question

ghalliday commented:
Duplicate the question, since it doesn't appear above:

I think you need to clarify the specification - whether a timer instance can be updated from multiple threads or not.

If the answer is yes, then the idea behind the Scope class doesn't work - because timers on one thread will prevent timers from other threads being updated.

If the answer is no (or that we don't care about small race conditions), then it is much more efficient to remove the critical section.

Whichever it is should be included as a comment before the class to make it clear.

Commit:
HPCC-15191 Extend TxSummary functionality to support timing milestones

Add timer class to track total elapsed time reported in one or more steps.

Update TxSummary to manage a collection of timer instances, referenced by name,
Extend the summary API to permit lookup of, and updates to, managed timers.
Add managed timer values to the serialized summary.

Extend the IEspContext interface to provide pass-through functions comparable
to its existing "trace summary" interface.

Signed-off-by: Tim Klemm <tim.klemm@lexisnexis.com>

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community_6.0.0-beta1.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

timothyklemm commented:
@ghalliday I removed the critical section from CumulativeTimer and added a comment stating that it is not thread safe, and individual instances are not to be used to time concurrent code blocks.

Commit:
Remove unnecessary CriticalBlock.

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community_6.0.0-beta1.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

rpastrana commented:
@richardkchapman based on Tim's latest changes, and offline conversation with Tim, this set of changes should now be addressing @ghalliday's concerns.
